### PR TITLE
Added support for AWS and specifying returnUri as config

### PR DESCRIPTION
**Pull request by remcobosman**
Adding so flexibility for setting the 'Return Uri' when the default window.location is not working. a simple trailing slash may cause failure when requesting a token.

Added the possibility to use 'application/x-www-form-urlencoded' in stead of 'application/json'. Aws requires this for correct payload, see: https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html

And added the 'RedirectUri' to the fetchToken payload, also required by Aws, but I do not know if this breaks it for non Aws token services.

I have never done anything in JavaScript, I have C# as a background. So let me know if I broke any rules.

Regards,

Remco Bosman

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -4,12 +4,15 @@ import createAuthContext from './lib/createAuthContext'
 const clientId = process.env.REACT_APP_CLIENT_ID || "8cb4904ae5581ecc2b3a1774"
 const clientSecret = process.env.REACT_APP_CLIENT_SECRET || "b683283462070edbac15a8fdab751ada0f501ab48a5f06aa20aee3be24eac9cc"
 const provider = process.env.REACT_APP_PROVIDER || "https://authenticate.u5auth.com"
+const redirectUri = process.env.REACT_RETURN_URI || "https://localhost"
+const useUrlEncodedForm = process.env.REACT_USE_URLENCODEDFORM || true
 
 const {AuthContext, Authenticated, useToken} = createAuthContext({
   clientId,
-  clientSecret,
```
Comment (Chris927) [inline src/App.js:10]: Should not be removed, the example oauth2 provider needs it.

File: src/App.js [hunk 2]
```diff
@@ -4,12 +4,15 @@ import createAuthContext from './lib/createAuthContext'
...
   provider,
-  // tokenEndpoint: 'http://localhost:3020/token' // If token endpoint is not "provider + '/token'"
```
Comment (Chris927) [inline src/App.js:12]: This should not be removed either, as it is documentation for those that need to customise the token endpoint.

File: src/lib/createAuthContext.js [hunk 3]
```diff
@@ -1,92 +1,195 @@
...
-
-  const context = createContext({})
-  const {Provider} = context
-
-  class Authenticated extends React.Component {
-    static contextType = context
-    componentDidMount() {
-      const { ensureAuthenticated } = this.context
-      ensureAuthenticated()
+"use strict";
```
Comment (Chris927) [inline src/lib/createAuthContext.js:1]: It seems you transpiled some source files and committed the transpiled version. This is a no-no: In a pull request, you must change as little as possible to achieve the desired change (whilst you can always improve existing code while making change). You must leave the coding style chosen by previous developers intact. Committing a transpiled file as source file reduces readability, breaks the coding style of the original, etc.

You can re-do your changes based on the original files, adhering to this feedback, then request me to merge again.

Also, please note that the `npm run build` is supposed to be used to transpile the source into the `./dist` folder, check `package.json` for details.
 
For now I'll have to reject your pull request, sorry.

File: src/App.js (hunk 1 above)
Comment (remcobosman) [reply src/App.js:10]: I was using the internal files I had used to create my own solution, but it got mixed up with the original. I brought everything back to its original state.

File: src/App.js (hunk 2 above)
Comment (remcobosman) [reply src/App.js:12]: I was using the internal files I had used to create my own solution, but it got mixed up with the original. I brought everything back to its original state.

File: src/lib/createAuthContext.js (hunk 3 above)
Comment (remcobosman) [reply src/lib/createAuthContext.js]: I was using the internal files I had used to create my own solution, but it got mixed up with the original. I brought everything back to its original state.

**Comment by remcobosman** [comment]
Can you please have another look.

File: src/App.js [hunk 4]
```diff
@@ -4,12 +4,16 @@ import createAuthContext from './lib/createAuthContext'
 const clientId = process.env.REACT_APP_CLIENT_ID || "8cb4904ae5581ecc2b3a1774"
 const clientSecret = process.env.REACT_APP_CLIENT_SECRET || "b683283462070edbac15a8fdab751ada0f501ab48a5f06aa20aee3be24eac9cc"
 const provider = process.env.REACT_APP_PROVIDER || "https://authenticate.u5auth.com"
+const redirectUri = process.env.REACT_RETURN_URI || "https://localhost"
```
Comment (Chris927) [inline src/App.js:7]: Why `https` (and not `http`) as default? You normally don't have / use https against localhost.

File: src/App.js [hunk 5]
```diff
@@ -4,12 +4,16 @@ import createAuthContext from './lib/createAuthContext'
...
+const useUrlEncodedForm = process.env.REACT_USE_URLENCODEDFORM || true
 
 const {AuthContext, Authenticated, useToken} = createAuthContext({
   clientId,
   clientSecret,
   provider,
   // tokenEndpoint: 'http://localhost:3020/token' // If token endpoint is not "provider + '/token'"
+  redirectUri,
```
Comment (Chris927) [inline src/App.js:15]: When I run the example (as per README.md), I get errors:

```
Failed to compile.

./src/lib/helpers/fetchToken.js
  Line 7:5:    'redirectUri' is not defined        no-undef
  Line 8:5:    'useUrlEncodedForm' is not defined  no-undef
  Line 14:14:  'useUrlEncodedForm' is not defined  no-undef
  Line 18:23:  'useUrlEncodedForm' is not defined  no-undef

Search for the keywords to learn more about each error.
```

Did you test the example? Please make sure the example works.

File: src/lib/helpers/fetchToken.js [hunk 6]
```diff
@@ -3,17 +3,22 @@ export const fetchToken = ({ clientId, clientSecret, code, verifier, tokenEndpoi
...
+  var body = useUrlEncodedForm ? Object.keys(payload).map(key => key + '=' + payload[key]).join('&') : JSON.stringify(payload);
+
   return fetch(tokenEndpoint, {
     headers: {
-      'Content-Type': 'application/json'
+      'Content-Type': useUrlEncodedForm ? 'application/x-www-form-urlencoded' : 'application/json'
     },
     method: 'POST',
-    body: JSON.stringify(payload)
+    body: JSON.stringify(body)
```
Comment (Chris927) [inline src/lib/helpers/fetchToken.js:21]: This looks wrong: If `useUrlEncodedForm` is `false`, then you stringify twice, and if `useUrlEncodedForm` is `true`, then the body should not be stringified, I suppose.

**Comment by remcobosman** [comment]
@Chris927, thanks for the reply! I think we don't need any changes at all at this moment. I can get the infra team on my end to configure the Return Uri properly, mitigating that problem. The problem was that this was defined with a trailing slash at the end, windows.location does not have that, but it is easily solved on the configuration end.

And when the encoded form `application/x-www-form-urlencoded` is the standard, that is fixed too. 

You can remove my pull-request, as it does not make much sense now.

Regards,

Remco Bosman

**Comment by Chris927** [comment]
Hi @remcobosman 

Thanks again!

There is a new version (0.9.10) published now, which uses the standard encoding.

Chris

**Pull request closed without merging**